File: src/integrator/MDIntegrator.cpp
/*
  Copyright (C) 2012,2013
      Max Planck Institute for Polymer Research
  Copyright (C) 2008,2009,2010,2011
      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
  
  This file is part of ESPResSo++.
  
  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
*/

#include <python.hpp>
#include "MDIntegrator.hpp"
#include "System.hpp"


namespace espressopp {
  namespace integrator {

    LOG4ESPP_LOGGER(MDIntegrator::theLogger, "MDIntegrator");

    //////////////////////////////////////////////////
    // Constructor              
    //////////////////////////////////////////////////

    MDIntegrator::MDIntegrator(shared_ptr<System> system) :
    SystemAccess(system)
    {
      LOG4ESPP_INFO(theLogger, "construct Integrator");
      if (!system->storage) {
        LOG4ESPP_ERROR(theLogger, "system has no storage");
      }
      timeFlag = true;
      step = 0;
      dt = 0.005;
    }
    
    MDIntegrator::~MDIntegrator()
    {
      LOG4ESPP_INFO(theLogger, "~Integrator");
    }
    
    void MDIntegrator::setTimeStep(real _dt)
    {
      if (_dt == 0.0) {
        System& system = getSystemRef();
        esutil::Error err(system.comm);
        std::stringstream msg;
        msg << "Timestep 'dt' must be non-zero!";
        err.setException(msg.str());
      }

      dt = _dt;
    }


    void MDIntegrator::addExtension(shared_ptr<integrator::Extension> extension) {
       //extension->setIntegrator(this); // this is done in python
       //std::cout << "type is: " << extension->type << "\n";

       /*
       //currently we do not check this, some Extensions are allowed multiple times,
       //e.g. ExtAnalyze or FixParticles with different particle groups
       // warn if there is already an extension of the same type
       for (ExtensionList::iterator it = exList.begin(); it != exList.end(); ++it) {
           if ((*it)->type == extension->type) {
              LOG4ESPP_WARN(theLogger, "extension of same type already added!");
              printf("\nWARNING: extension of same type already added!\n\n");
           }
       }
       */

       // add extension to the list
       exList.push_back(extension);
    }

    int MDIntegrator::getNumberOfExtensions() {
    	return exList.size();
    }

    shared_ptr<integrator::Extension> MDIntegrator::getExtension(int k) {
    	return exList[k];
    }

    //////////////////////////////////////////////////
    // REGISTRATION WITH PYTHON
    //////////////////////////////////////////////////

    void MDIntegrator::registerPython() {

      using namespace espressopp::python;

      // Note: use noncopyable and no_init for abstract classes
      class_<MDIntegrator, boost::noncopyable>
        ("integrator_MDIntegrator", no_init)
        .add_property("dt", &MDIntegrator::getTimeStep, &MDIntegrator::setTimeStep)
        .add_property("step", &MDIntegrator::getStep, &MDIntegrator::setStep)
        .add_property("system", &SystemAccess::getSystem)
        .def("run", &MDIntegrator::run)
        .def("addExtension", &MDIntegrator::addExtension)
        .def("getNumberOfExtensions", &MDIntegrator::getNumberOfExtensions)
        .def("getExtension", &MDIntegrator::getExtension)
        ;
    }
  }
}

File: src/integrator/MDIntegrator.py
#  Copyright (C) 2012,2013
#      Max Planck Institute for Polymer Research
#  Copyright (C) 2008,2009,2010,2011
#      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
#  
#  This file is part of ESPResSo++.
#  
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>. 


r"""
**********************************
espressopp.integrator.MDIntegrator
**********************************



.. function:: espressopp.integrator.MDIntegrator.addExtension(extension)

		:param extension: 
		:type extension: 
		:rtype: 

.. function:: espressopp.integrator.MDIntegrator.getExtension(k)

		:param k: 
		:type k: 
		:rtype: 

.. function:: espressopp.integrator.MDIntegrator.getNumberOfExtensions()

		:rtype: 

.. function:: espressopp.integrator.MDIntegrator.run(niter)

		:param niter: 
		:type niter: 
		:rtype: 
"""
from espressopp import pmi
from _espressopp import integrator_MDIntegrator

class MDIntegratorLocal(object):

    def run(self, niter):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            return self.cxxclass.run(self, niter)

    def addExtension(self, extension):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            
            # set integrator and connect to it
            extension.cxxclass.setIntegrator(extension, self)
            extension.cxxclass.connect(extension)
            
            return self.cxxclass.addExtension(self, extension)
        
    def getExtension(self, k):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            return self.cxxclass.getExtension(self, k)

    def getNumberOfExtensions(self):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            return self.cxxclass.getNumberOfExtensions(self)

if pmi.isController :
    class MDIntegrator(object):

        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            pmiproperty = [ 'dt', 'step' ],
            pmicall = [ 'run', 'addExtension', 'getExtension', 'getNumberOfExtensions' ]
            )
